### DISCUSSION TITLE
AlpineJS morph plugin problems & replacing `innerHTML` ?

### DISCUSSION THREAD

**Discussion by DEV-ONI** [alpinejs/alpine#3120] [1. Help]
I love using Alpine and have started using it for Shopify projects for clients, 
however the morph plugin seems to have problems / inconsistencies in its behavior (?)

My use-case is with the [Shopify Section Rendering API](https://shopify.dev/api/section-rendering), which returns the HTML of a requested "Shopify section", and `Alpine.morph` can be called on the sections to update their content. **_This usually works, but_** there are instances wherein "morphing" a section causes the layout to break:
- some elements being inserted into parents where they don't belong
- some elements being replaced incorrectly
- `patchAttributes` raising an error when `x-on` shorthand is used due to `updateAttribute`
- etc...

Most of these problems can be mitigated by use of the `key` attribute or hooks, however it is difficult to predict the behavior of `Alpine.morph` on different sections where the markup can change. I've also checked if this was by any chance caused by invalid HTML to no avail.

---

I think an alternative would be to simply replace the `innerHTML` on the Alpine components? 
Are there any caveats to this? 

---

Also I'm currently compiling some of the bugs I've found and hope to add an issue / contribute soon.
And thank you for the great library!

**Comment by Tim-Wils** [comment]
Depending on the usecase, you could probably do an el.innerHTML replacement.
You could additionally pause Alpine's watcher for a moment while replacing your html by calling Alpine.mutateDom(...callback with replace here...)
Like so:
```
Alpine.mutateDom(() => {
el.innerHTML = newHTML;
});
```

**Reply by DEV-ONI** [reply]
Thanks @Tim-Wils.

Forgive my ignorance, but would this cause Alpine to not preserve the component?
I'm not too familiar with the internals of AlpineJS, but from this discussion it seems to be the case? - #2789 

Am I right in saying if I were to replace the innerHTML as opposed to "morphing", the significant difference is it would re-initialize any components within, right?

**Reply by Tim-Wils** [reply]
You’re right that if you want replace x-data components and keep the state that you’ll have to use Alpine.morph :) I was just offering a possible alternative solution. 
Could you maybe provide a minimal example of the morph-problem you bumped into in a codepen, so that we could take a look at the bugs?

**Reply by DEV-ONI** [reply]
Okay sure. I'll try to find out what's causing the problem and replicate it.
